File: web/src/constants/api-endpoints.ts
/**
 * API Endpoints constants for frontend
 * Frontend can easily use these constants to construct API URLs
 */

const API_BASE = '/api';

export const API_ENDPOINTS = {
  auth: {
    login: `${API_BASE}/auth/login`,
    register: `${API_BASE}/auth/register`,
    verifyEmail: `${API_BASE}/auth/verify-email`,
    resendVerification: `${API_BASE}/auth/resend-verification`,
    refresh: `${API_BASE}/auth/refresh`,
    logout: `${API_BASE}/auth/logout`,
    logoutAll: `${API_BASE}/auth/logout-all`,
    forgotPassword: `${API_BASE}/auth/forgot-password`,
    resetPassword: `${API_BASE}/auth/reset-password`,
  },

  users: {
    getMe: `${API_BASE}/users/me`,
    updateMe: `${API_BASE}/users/me`,
    changePassword: `${API_BASE}/users/me/change-password`,
  },

  lockers: {
    getAll: `${API_BASE}/lockers`,
    create: `${API_BASE}/lockers`,
    getById: (id: string) => `${API_BASE}/lockers/${id}`,
    update: (id: string) => `${API_BASE}/lockers/${id}`,
    delete: (id: string) => `${API_BASE}/lockers/${id}`,
    getAvailable: `${API_BASE}/lockers/available`,
    updateSlotStatus: (id: string, slotIndex: number) =>
      `${API_BASE}/lockers/${id}/slots/${slotIndex}/status`,
  },

  bookings: {
    getById: (id: string) => `${API_BASE}/bookings/${id}`,
    getMy: `${API_BASE}/bookings/my`,
    create: `${API_BASE}/bookings`,
    setPin: (id: string) => `${API_BASE}/bookings/${id}/set-pin`,
    verifyPin: (id: string) => `${API_BASE}/bookings/${id}/verify-pin`,
    complete: (id: string) => `${API_BASE}/bookings/${id}/complete`,
    cancel: (id: string) => `${API_BASE}/bookings/${id}/cancel`,
  },

  packages: {
    getAll: `${API_BASE}/packages`,
    create: `${API_BASE}/packages`,
    getById: (id: string) => `${API_BASE}/packages/${id}`,
    update: (id: string) => `${API_BASE}/packages/${id}`,
    delete: (id: string) => `${API_BASE}/packages/${id}`,
  },

  payments: {
    getById: (id: string) => `${API_BASE}/payments/${id}`,
    getByBookingId: (bookingId: string) => `${API_BASE}/payments/booking/${bookingId}`,
    getMy: `${API_BASE}/payments/my`,
    create: `${API_BASE}/payments`,
    complete: (id: string) => `${API_BASE}/payments/${id}/complete`,
  },

  admin: {
    users: {
      getAll: `${API_BASE}/admin/users`,
      updateRole: (id: string) => `${API_BASE}/admin/users/${id}/role`,
      deactivate: (id: string) => `${API_BASE}/admin/users/${id}/deactivate`,
      activate: (id: string) => `${API_BASE}/admin/users/${id}/activate`,
    },
    bookings: {
      getAll: `${API_BASE}/admin/bookings`,
    },
    payments: {
      getAll: `${API_BASE}/admin/payments`,
    },
  },

  health: {
    check: `${API_BASE}/health`,
  },
};

// Helper function to build URL with query parameters
export const buildUrl = (
  endpoint: string,
  params?: Record<string, string | number | boolean>
): string => {
  if (!params || Object.keys(params).length === 0) {
    return endpoint;
  }

  const searchParams = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    searchParams.append(key, String(value));
  });

  return `${endpoint}?${searchParams.toString()}`;
};

// Example usage in frontend:
// import { API_ENDPOINTS, buildUrl } from '@/constants/api-endpoints'
//
// GET request:
// fetch(API_ENDPOINTS.users.getMe)
//
// With ID parameter:
// fetch(API_ENDPOINTS.lockers.getById('locker-123'))
//
// With query parameters:
// fetch(buildUrl(API_ENDPOINTS.bookings.getMy, { status: 'Active' }))
//
// POST request:
// fetch(API_ENDPOINTS.auth.login, {
//   method: 'POST',
//   headers: { 'Content-Type': 'application/json' },
//   body: JSON.stringify(loginData)
// })

File: backend/src/Locker.Backend.Application/ApiEndpoints.cs
namespace Locker.Backend.Application;

/// <summary>
/// Centralized API endpoint constants for easy frontend integration.
/// Use these constants to construct API URLs across the application.
/// </summary>
public static class ApiEndpoints
{
    private const string ApiBase = "/api";

    public static class Auth
    {
        private const string Base = $"{ApiBase}/auth";
        public const string Login = $"{Base}/login";
        public const string Register = $"{Base}/register";
        public const string VerifyEmail = $"{Base}/verify-email";
        public const string ResendVerification = $"{Base}/resend-verification";
        public const string Refresh = $"{Base}/refresh";
        public const string Logout = $"{Base}/logout";
        public const string LogoutAll = $"{Base}/logout-all";
        public const string ForgotPassword = $"{Base}/forgot-password";
        public const string ResetPassword = $"{Base}/reset-password";
    }

    public static class Users
    {
        private const string Base = $"{ApiBase}/users";
        public const string GetMe = $"{Base}/me";
        public const string UpdateMe = $"{Base}/me";
        public const string ChangePassword = $"{Base}/me/change-password";
    }

    public static class Lockers
    {
        private const string Base = $"{ApiBase}/lockers";
        public const string GetAll = Base;
        public const string Create = Base;
        public const string GetById = $"{Base}/{{id}}";
        public const string Update = $"{Base}/{{id}}";
        public const string Delete = $"{Base}/{{id}}";
        public const string GetAvailable = $"{Base}/available";
        public const string UpdateSlotStatus = $"{Base}/{{id}}/slots/{{slotIndex}}/status";
    }

    public static class Bookings
    {
        private const string Base = $"{ApiBase}/bookings";
        public const string GetById = $"{Base}/{{id}}";
        public const string GetMy = $"{Base}/my";
        public const string Create = Base;
        public const string SetPin = $"{Base}/{{id}}/set-pin";
        public const string VerifyPin = $"{Base}/{{id}}/verify-pin";
        public const string Complete = $"{Base}/{{id}}/complete";
        public const string Cancel = $"{Base}/{{id}}/cancel";
    }

    public static class Packages
    {
        private const string Base = $"{ApiBase}/packages";
        public const string GetAll = Base;
        public const string Create = Base;
        public const string GetById = $"{Base}/{{id}}";
        public const string Update = $"{Base}/{{id}}";
        public const string Delete = $"{Base}/{{id}}";
    }

    public static class Orders
    {
        private const string Base = $"{ApiBase}/orders";
        public const string GetById = $"{Base}/{{id}}";
        public const string GetMy = $"{Base}/my";
        public const string Reserve = $"{Base}/reserve";
        public const string Confirm = $"{Base}/{{id}}/confirm";
        public const string SetPin = $"{Base}/{{id}}/set-pin";
        public const string Activate = $"{Base}/{{id}}/activate";
        public const string Complete = $"{Base}/{{id}}/complete";
        public const string Cancel = $"{Base}/{{id}}/cancel";
        public const string Extend = $"{Base}/{{id}}/extend";
        public const string GetAvailableSlots = $"{Base}/availability/slots";
        public const string LinkPayment = $"{Base}/{{id}}/payment";
    }

    public static class Payments
    {
        private const string Base = $"{ApiBase}/payments";
        public const string GetById = $"{Base}/{{id}}";
        public const string GetByBookingId = $"{Base}/booking/{{bookingId}}";
        public const string GetMy = $"{Base}/my";
        public const string Create = Base;
        public const string Complete = $"{Base}/{{id}}/complete";
    }

    public static class Admin
    {
        private const string Base = $"{ApiBase}/admin";

        public static class Users
        {
            private const string BaseUsers = $"{Base}/users";
            public const string GetAll = BaseUsers;
            public const string UpdateRole = $"{BaseUsers}/{{id}}/role";
            public const string Deactivate = $"{BaseUsers}/{{id}}/deactivate";
            public const string Activate = $"{BaseUsers}/{{id}}/activate";
        }

        public static class Bookings
        {
            private const string BaseBookings = $"{Base}/bookings";
            public const string GetAll = BaseBookings;
        }

        public static class Payments
        {
            private const string BasePayments = $"{Base}/payments";
            public const string GetAll = BasePayments;
        }
    }

    public static class Health
    {
        public const string Check = $"{ApiBase}/health";
    }
}
